File: app/src/main/java/com/zack/enderweather/adapter/CityAdapter.java
package com.zack.enderweather.adapter;

import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.zack.enderweather.R;
import com.zack.enderweather.bean.Weather;

import java.util.List;
import java.util.Locale;

public class CityAdapter extends RecyclerView.Adapter<CityAdapter.ViewHolder> {

    private List<Weather> weatherList;

    private OnCityItemClickListener onCityItemClickListener;
    private OnUpdateButtonClickListener onUpdateButtonClickListener;
    private OnDeleteButtonClickListener onDeleteButtonClickListener;

    public CityAdapter(List<Weather> weatherList) {
        this.weatherList = weatherList;
    }

    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View itemView = LayoutInflater.from(parent.getContext()).inflate(R.layout.item_city, parent, false);
        return new ViewHolder(itemView);
    }

    @Override
    public void onBindViewHolder(final ViewHolder holder, int position) {
        Weather weather = weatherList.get(position);

        holder.cityText.setText(weather.getBasicInfo().getCityName());
        holder.weatherText.setText(weather.getBasicInfo().getUpdateTime().isEmpty() ? "--°C | --" :
                String.format("%s°C | %s", weather.getCurrentInfo().getTemperature(),
                        weather.getCurrentInfo().getCondition()));

        if (onCityItemClickListener != null) {
            holder.itemView.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    onCityItemClickListener.onCityItemClick(holder.getLayoutPosition());
                }
            });
        }
        if (onUpdateButtonClickListener != null) {
            holder.updateButton.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    onUpdateButtonClickListener.onUpdateButtonClick(holder.getLayoutPosition());
                }
            });
        }
        if (onDeleteButtonClickListener != null) {
            holder.deleteButton.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    onDeleteButtonClickListener.onDeleteButtonClick(holder.getLayoutPosition());
                }
            });
        }
    }

    @Override
    public int getItemCount() {
        return weatherList.size();
    }

    class ViewHolder extends RecyclerView.ViewHolder {
        TextView cityText, weatherText;
        ImageView updateButton, deleteButton;

        public ViewHolder(View itemView) {
            super(itemView);
            cityText = (TextView) itemView.findViewById(R.id.text_city);
            weatherText = (TextView) itemView.findViewById(R.id.text_weather);
            updateButton = (ImageView) itemView.findViewById(R.id.btn_update);
            deleteButton = (ImageView) itemView.findViewById(R.id.btn_delete);
        }
    }

    public interface OnCityItemClickListener {
        void onCityItemClick(int position);
    }

    public void setOnCityItemClickListener(OnCityItemClickListener listener) {
        onCityItemClickListener = listener;
    }

    public interface OnUpdateButtonClickListener {
        void onUpdateButtonClick(int position);
    }

    public void setOnUpdateButtonClickListener(OnUpdateButtonClickListener listener) {
        onUpdateButtonClickListener = listener;
    }

    public interface OnDeleteButtonClickListener {
        void onDeleteButtonClick(int position);
    }

    public void setOnDeleteButtonClickListener(OnDeleteButtonClickListener listener) {
        onDeleteButtonClickListener = listener;
    }
}

File: app/src/main/java/me/imzack/app/cold/view/adapter/CityAdapter.kt
package me.imzack.app.cold.view.adapter

import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import kotlinx.android.extensions.LayoutContainer
import kotlinx.android.synthetic.main.item_list_city.*

import me.imzack.app.cold.R
import me.imzack.app.cold.common.Constant
import me.imzack.app.cold.model.DataManager
import me.imzack.app.cold.model.bean.Weather
import me.imzack.app.cold.util.ResourceUtil

class CityAdapter : RecyclerView.Adapter<CityAdapter.ItemViewHolder>() {

    var onCityItemClickListener: ((position: Int) -> Unit)? = null
    var onUpdateButtonClickListener: ((position: Int) -> Unit)? = null
    var onDeleteButtonClickListener: ((position: Int) -> Unit)? = null

    private val updateAnim = ResourceUtil.getAnimation(R.anim.anim_update)

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int) =
            ItemViewHolder(LayoutInflater.from(parent.context).inflate(R.layout.item_list_city, parent, false))

    override fun onBindViewHolder(holder: ItemViewHolder, position: Int) {
        val (_, cityName, current, _, _, updateTime, status) = DataManager.getWeather(position)

        holder.vCityNameText.text = cityName
        holder.vWeatherText.text = if (updateTime == 0L) "${Constant.UNKNOWN_DATA} | ${Constant.UNKNOWN_DATA}" else "${current.temperature} | ${DataManager.getConditionByCode(current.conditionCode)}"

        if (status == Weather.STATUS_ON_UPDATING) {
            holder.vUpdateButton.startAnimation(updateAnim)
        }

        holder.itemView.setOnClickListener { onCityItemClickListener?.invoke(holder.layoutPosition) }
        holder.vUpdateButton.setOnClickListener { onUpdateButtonClickListener?.invoke(holder.layoutPosition) }
        holder.vDeleteButton.setOnClickListener { onDeleteButtonClickListener?.invoke(holder.layoutPosition) }
    }

    override fun getItemCount() = DataManager.cityCount

    class ItemViewHolder(override val containerView: View) : RecyclerView.ViewHolder(containerView), LayoutContainer
}
